SDL/SDLJoystick.cpp

```diff
@@ -11,167 +11,209 @@ extern "C" {
 	}
 }
 
-SDLJoystick::SDLJoystick(bool init_SDL ): thread(NULL), running(true) {
+SDLJoystick::SDLJoystick(bool init_SDL ): thread(NULL), running(true), player1Controller(NULL), player1JoyInstanceID(-1) {
 	if (init_SDL)
 	{
-		SDL_setenv("SDL_JOYSTICK_ALLOW_BACKGROUND_EVENTS", "1", 0);
-		SDL_Init(SDL_INIT_JOYSTICK | SDL_INIT_VIDEO);
+		SDL_SetHint(SDL_HINT_JOYSTICK_ALLOW_BACKGROUND_EVENTS, "1");
+		SDL_Init(SDL_INIT_JOYSTICK | SDL_INIT_VIDEO | SDL_INIT_GAMECONTROLLER);
 	}
+	
+	if (SDL_GameControllerAddMappingsFromFile("assets/gamecontrollerdb.txt") == -1)
+	{
+			printf("Failed to load control pad mappings, please place gamecontrollerdb.txt in your assets directory\n");
+	}
+	else {
+		printf("Loaded game controller mappings for SDL2\n");
+		setUpPlayer1();
+	}
+}
 
+void SDLJoystick::setUpPlayer1()
+{
 	int numjoys = SDL_NumJoysticks();
-	SDL_JoystickEventState(SDL_ENABLE);
+	int joyIndex = -1;
 	for (int i = 0; i < numjoys; i++) {
-		joys.push_back(SDL_JoystickOpen(i));
-//		printf("Initialized joystick %d: %s",i,SDL_JoystickNameForIndex(i));
-		if(strstr(SDL_JoystickNameForIndex(i),"PLAYSTATION(R)3 Controller"))
-			g_Config.bPS3Controller = true;
+		if (joyIndex == -1 && SDL_IsGameController(i))
+		{
+			joyIndex = i;
+			break;
+		}
+	}
+	if (joyIndex != -1)
```

Comment:
Well, the UI has a way to map buttons.  It won't work inside the UI for navigating the UI (otherwise you could trap yourself), but if the game exposes a typical hat the old mapping can be usable.  It might be better than not being able to use the controller at all.

-[Unknown]